### ISSUE TITLE
Breaking the Softmax Bottleneck: A High-Rank RNN Language Model

### ISSUE THREAD

**Issue by noowad93** [noowad93/ml-paper-summaries#13]
# 논문 정보 (URL, 저자 등)

[https://openreview.net/pdf?id=HkwZSG-CZ](https://openreview.net/pdf?id=HkwZSG-CZ)

Carnegie Mellon University
(XL-Net 저자)
ICLR 2018

# 개요

- 일반적으로 언어 모델을 모델링할 때는 RNN으로부터 계산된 hidden vector와 word embedding을 dot-product하고 (logit), 이것에 softmax를 취한 값을 각 토큰이 등장할 확률로 간주하여 모델링하는 경우가 많습니다.
- 본 논문에서는 이렇게 RNN+Softmax를 이용하여 Natural Language를 모델링하는 방식의 문제점을 지적하고 (Softmax Bottleneck), 이 문제를 간단하게 해결할 수 있는 Mixture of Softmax (MoS) 구조를 제안합니다.
    - MoS 구조를 이용하여 다양한 언어 모델 데이터셋에 대해 SOTA를 달성했습니다.

# Softmax Bottleneck

**언어 모델**

- 일반적으로 언어 모델은 아래와 같은 식처럼 주어진 이전 컨텍스트에 따른 다음 토큰의 등장 확률로 모델링합니다.

    
   <img width="465" alt="_2020-05-26__5 32 40" src="https://user-images.githubusercontent.com/49383323/93034207-49062480-f674-11ea-937d-94e640117a68.png">

- 그리고, 위의 확률값을 모델링하는 구조로는 RNN+Softmax 구조를 주로 이용합니다. RNN으로부터 얻은 Context의 fixed-length hidden vector (RNN 각 스텝의 output)와, Output word embedding matrix의 dot-product 값을 logit으로 사용하고, 이 값에 softmax를 취해 확률 값을 얻습니다.

    
   <img width="189" alt="_2020-05-26__5 37 32" src="https://user-images.githubusercontent.com/49383323/93034646-b9fa0c00-f675-11ea-86fc-184fb38b1b15.png">

- 이 모델은 우리가 실제로 알고 싶은 True Language Model Probability를 잘 모방할 수 있는 파라미터를 학습하게 됩니다.
- 본 논문에서는 "과연 RNN+Softmax 구조로 학습한 파라미터가 실제 확률 값을 잘 모방할 수 있을까?"에 의문을 제기합니다.

**Language Modeling as Matrix Factorization**

- 우선, 문제를 수학적으로 정식화하기 위해서 가능한 컨텍스트의 수를 N개의 원소를 갖는 유한한 집합으로 가정합니다.

    <img width="286" alt="_2020-05-26__5 42 50" src="https://user-images.githubusercontent.com/49383323/93034806-486e8d80-f676-11ea-89ca-1efb67bb38cb.png">

- 여기에 더하여, H, W, A라는 3개의 매트릭스를 정의합니다.
    
   <img width="625" alt="_2020-05-26__7 40 25" src="https://user-images.githubusercontent.com/49383323/93034816-502e3200-f676-11ea-8777-19281200ec64.png">

    - H는 N*d 차원의 매트릭스로, Context Vector에 대한 hidden vector
    - W는 M*d 차원의 매트릭스로, 워드 임베딩 (M은 어휘 사이즈)
    - A는 N*M 차원의 매트릭스로, 실제 (True) 데이터 분포의 log probability
- 우리가 RNN+Softmax로 언어 모델링을 한다는 것을 실제 분포 A를 H와 W로 factorization하여 푸는 것이라 간주하자!
- 모델이 예측하는 log token probability (logits)는 아래 식과 같이 구할 수 있습니다.
    
    <img width="664" alt="_2020-05-26__7 51 29" src="https://user-images.githubusercontent.com/49383323/93034886-8075d080-f676-11ea-9c4a-7cbf7efed1a3.png">

- A에 row-wise shift를 적용한 Set of Matrices F(A)를 아래와 같이 정의합니다. (Softmax는 Softmax(WX)=Softmax(WX+b) (b는 스칼라 값)라는 성질이 있기 때문에)
    
   <img width="378" alt="_2020-05-26__8 10 15" src="https://user-images.githubusercontent.com/49383323/93035063-f2e6b080-f676-11ea-8b9c-ba52def7306e.png">

    - 단, J는 행렬의 모든 값이 1인 행렬이기 때문에 우변에 더해진 값은 아래와 같은 형태입니다.

    <img width="506" alt="_2020-05-26__8 12 07" src="https://user-images.githubusercontent.com/49383323/93035010-d2b6f180-f676-11ea-8b21-85e78f0b4ec1.png">

- 이 F(A)는 아래와 같은 두 가지 속성을 갖습니다.

    <img width="585" alt="_2020-05-26__8 15 28" src="https://user-images.githubusercontent.com/49383323/93035025-d8143c00-f676-11ea-8a8d-39fa60ca8252.png">

    1. F(A)는 True data distribution에 대응하는 모든 가능한 logits를 정의한 집합이다.
    2. F(A)에 포함된 matrix들의 Rank의 차이는 1 이하이다.
    - 증명은 페이퍼 맨 뒤에...
- 결국 궁금한건, **모델이 예측하는 logits 값이 F(A) 즉, true token probability의 집합에 속할 수 있는가?** 입니다.
    - 즉, **아래의 명제가 성립하는가** 입니다.

    <img width="546" alt="_2020-05-26__8 44 57" src="https://user-images.githubusercontent.com/49383323/93035119-1c9fd780-f677-11ea-8a81-493603f304f8.png">

    - 속할 수 있다면, 모델의 예측 값은 실제 데이터 분포를 완전히 재현해낼 수 있다는 말이 되는 것이고, 그렇지 않다면, 이 구조를 갖는 모델의 예측 값은 세상의 모든 데이터를 세상의 모든 GPU로 무한히 학습시켜도 실제 데이터 분포를 완전히 재현해낼수 없다는 것.
- 이제 여기서 드러나는 문제가, 위의 명제에서 좌변의 Rank는 반드시 embedding_size d보다 작은 값일 수 밖에 없기 때문에, 우변의 Rank가 d보다 크다면, 좌변과 같은 matrix factorization 방법으로는 절대로 우변을 복구할 수 없다는 점입니다.
    - 단, 주의해야 할 점은, 이 문제의 원인은 H의 학습 구조로 RNN을 사용해서가 아니고, prob을 구할 때 Softmax를 사용해서 입니다. (RNN이 요새 잘 안쓰이긴 하지만, 기본적으로 universal approximator라 문제 없습니다.)
    - 이 문제를 이 논문에서는 **Softmax Bottleneck**이라고 말하고 있습니다.

**가설: 자연어는 High-Rank인가?**

- 위에서 지적한 Rank 문제는 사실, True probability가 low-Rank라면 큰 문제가 없습니다 (Rank가 d이하라면).
    - 따라서, 과연 **언어 모델링 태스크에서 rank(A)의 값은 정말로 큰 값을 가질까?** 라는 의문을 제기할 수 있습니다.
- 하지만, 실제 분포 A는 이론상 구할 수 없는 값이기 때문에 엄밀히 증명을 할 수는 없지만, 아래와 같은 관찰을 통해 자연어가 High Rank라는 가설을 세울 수는 있습니다.
    1. 자연어는 굉장히 Context-dependent하다. 예를 들어, 국제 정치 기사에서 north라는 단어가 등장하면, 다음 단어로 korea 나 korean 등의 확률이 높겠지만, 일반 교과서에서는 그렇지 않을 것입니다. 이런 미묘한 문맥 의존성은 high-rank의 A에서 나타나는 특징이라고 할 수 있습니다.
    2. 만약 A가 low-rank라 가정한다면 (수백 정도), 모든 semantic meanings가 몇백개의 basis meaning의 선형합으로 나타내어 질 수 있다는 것인데... 절대로 그렇지 않습니다.
    3. 경험적으로, 본 논문에서 high-rank한 언어 모델을 학습해보니 훨씬 성능이 좋았습니다.
- 위의 명제에서 좌변의 Rank는 D(임베딩 사이즈) 이하의 값을 갖고, 우변의 Rank는 M (어휘 사이즈) 이하의 값을 갖는데, 일반적으로 D << M이기 때문에 자연어가 D보다 큰 High-Rank라면, 문제가 생길 수 밖에 없는 구조입니다.

## Mixture of Softmax

- 본 논문에서 위의 Softmax Bottleneck을 해결하기 위해 제안하는 방법입니다.
- 이 방법은 간단히 말하자면, Hidden vector h를 만들 때 Mixture 구조로 복수개의 h를 만들고, 해당 h를 선택할 prior (phi) 또한 학습하여 아래 식과 같이 모델링하는 방법입니다.

    
<img width="361" alt="_2020-05-26__9 34 09" src="https://user-images.githubusercontent.com/49383323/93035190-4e18a300-f677-11ea-9c90-e6f272eb25ff.png">

- 이렇게 모델링을 하면, 모델의 logit이 log_sum_exp형태이기 때문에 비선형 함수가 되기 때문에, Rank가 d 이하라는 제약이 사라집니다.
- 또한, 위에서 제시한 MoS 모델이 단순히 **Mixture 구조** 때문이 아니라 **Softmax Bottleneck** 문제를 해결했다라고 주장하기 위해, 아래와 같이 Context를 Mixture한 구조도 같이 제시합니다.
    
   <img width="288" alt="_2020-05-26__9 37 54" src="https://user-images.githubusercontent.com/49383323/93035212-5ec91900-f677-11ea-88d7-b9fc6a69369a.png">

    - 그래서, 이 구조보다 MoS가 더 좋은 성능을 낸다면, **MoS가 Softmax Bottleneck 문제를 해결했기 때문에 성능이 좋아졌다** 라고 주장할 수 있게 됩니다.

# 실험

- 다양한 언어 모델링 데이터셋을 이용하여 모델의 우수성을 검증했습니다.
    
   <img width="644" alt="_2020-05-26__9 39 11" src="https://user-images.githubusercontent.com/49383323/93035343-af407680-f677-11ea-8d50-0babf3f75163.png">

- Seq2Seq 태스크에 대해서도 마찬가지로 우수한 성능을 보여줍니다.
    
   <img width="616" alt="_2020-05-26__9 40 13" src="https://user-images.githubusercontent.com/49383323/93035351-b6678480-f677-11ea-8bd5-99c1678c645e.png">

- MoC보다 MoS가 더 성능이 우수하다는 걸 확인했습니다.
    
   <img width="624" alt="_2020-05-26__9 40 42" src="https://user-images.githubusercontent.com/49383323/93035359-ba93a200-f677-11ea-9709-89bcbe4ff5f3.png">

## Ablation Study

- 실제로 MoS가 high-rank probability matrix를 학습할 수 있다는 것을 보이기 위해, SVD를 통해 실제 Rank를 계산했습니다.
    
   <img width="500" alt="_2020-05-26__9 41 46" src="https://user-images.githubusercontent.com/49383323/93035389-d0a16280-f677-11ea-94a9-87048b6f03e5.png">

- 또한, Mixture의 개수를 늘리면 Log-probability matrix의 Rank가 올라가고, Perplexity가 증가하는 경향을 보입니다.
    
    <img width="541" alt="_2020-05-26__9 42 51" src="https://user-images.githubusercontent.com/49383323/93035399-d5fead00-f677-11ea-81d5-9c4291e5011a.png">

- 한편, Rank가 큰 문제가 되지 않는 Character-level 언어 모델링 태스크에서는 vocab size가 hidden size보다 작기 때문에 log-probability matrix의 Rank가 V가 아닌 D에 의해 제약이 걸립니다. 이 경우에는 MoS가 큰 성능 향상을 보이지 않습니다.
    
    <img width="606" alt="_2020-05-26__9 44 57" src="https://user-images.githubusercontent.com/49383323/93035406-dac36100-f677-11ea-8e14-c7d96a21362b.png">

- MoS의 한가지 단점은, 무거운 Softmax 연산을 여러번 수행해야 한다는 점입니다.
    - 다만, Softmax 연산은 쉽게 병렬화 가능하기 때문에, 멀티 GPU 환경에서는 학습 시간이 sub-linear하다고 합니다.

# 논의할만한 내용

- 일반적으로 Transformer를 최적화할 때 보통, 레이어 개수나 헤드 개수 등을 중요하게 생각하는데, 위 논문에 의하면 Softmax를 써서 확률을 표현하는 모델링을 한다고 했을 때, 모델이 표현할 수 있는 prob의 Rank는 embedding_size에 제한됨.
    - 따라서, embedding_size 역시 언어 모델에서는 중요한 하이퍼파라미터이고, 최근의 성공적인 생성 모델들 (Meena, Blender) 등도 큰 hidden_size의 영향이 있지 않았을까 생각함.
